shell-integration/fish/vendor_conf.d/kitty-shell-integration.fish

```diff
@@ -20,66 +22,177 @@ function __ksi_schedule --on-event fish_prompt -d "Setup kitty integration after
+    #     block, underscore (underline), line (bar, beam)
+    #
+    # To make the cursor blink, add `blink` at the end.
+    #
+    # Example:
+    #     set -g kitty_fish_prompt_cursor line blink
+    #
     if not contains "no-cursor" $_ksi
         and not functions -q __ksi_set_cursor
 
-        function __ksi_set_cursor --on-variable fish_key_bindings -d "Set cursor shape for fish default mode"
-            if test "$fish_key_bindings" = fish_default_key_bindings
-                not functions -q __ksi_bar_cursor __ksi_block_cursor || return
+        # Set cursor shape escape sequence `ESC [ Ps SPACE q`
+        # Ps:
+        # blinking: block 1, underline 3, bar 5
+        # steady:   block 2, underline 4, bar 6
+        function __ksi_init_cursor_var -d "Initialize cursor shape CSI escape sequence"
+            set --local varname kitty_fish_$argv[1]_cursor
+            set --local val 0
+            if contains $argv[1] default insert replace_one visual
+                and test -n "fish_$argv[1]_cursor"
+                set varname fish_$argv[1]_cursor
+            end
+            if test -n "$$varname"
+                set --local shape (string split " " -- "$$varname")
+                switch $shape[1]
+                    case block
+                        set val 2
+                    case underline underscore
+                        set val 4
+                    case bar beam line
+                        set val 6
+                end
+                test $val -gt 0 && test "$shape[2]" = blink
+                and set val (math $val - 1)
+                set --global __ksi_cursor_$argv[1] "\e[$val q"
+            end
+            test $val -gt 0
+            or set --global __ksi_cursor_$argv[1] "\e[$argv[2] q"
+        end
 
-                function __ksi_bar_cursor --on-event fish_prompt
-                    printf "\e[5 q"
+        function __ksi_init_vi_cursor -d "Initialize the vi cursor shapes"
+            # Set default cursor shapes
+            # Use bar in insert mode, underline in replace mode, block in default and visual mode,
+            # all with a blinking cursor.
+            set --local _modes default insert replace_one visual
+            set --local _shapes 1 5 3 1
+            for i in 1 2 3 4
+                __ksi_init_cursor_var $_modes[$i] $_shapes[$i]
+            end
+
+            # The cursor shapes will be frozen after initialization and will not be read during runtime
+            printf "
+                function __ksi_vi_prompt_cursor -d 'Set cursor shape for current vi mode'
+                    switch \$fish_bind_mode
+                        case default; printf \"%s\"
+                        case insert; printf \"%s\"
+                        case replace_one; printf \"%s\"
+                        case visual; printf \"%s\"
+                    end
                 end
+            " $__ksi_cursor_default $__ksi_cursor_insert $__ksi_cursor_replace_one $__ksi_cursor_visual | source
```

Comment:
Why is this function defined via `source`?

Reply:
All these variables will be cleaned up after initialization, and no more steps such as switch (for each shape) will be performed to recalculate them at each repaint.
Maybe this is a bad practice?